src/Tuura/Plato/Translation.hs

```diff
@@ -76,8 +78,30 @@ validateInterface signs circuit
   where
     unused       = filter ((==Unused) . interface circuit) signs
 
-cartesianProduct :: NonEmpty.NonEmpty [a] -> [[a]]
-cartesianProduct l = sequence (NonEmpty.toList l)
+cartesianProduct :: Ord a => NonEmpty.NonEmpty [a] -> [[a]]
+cartesianProduct l = removeSupersets sortByLength
+  where
+    sequenced    = sequence (NonEmpty.toList l)
+    removeDupes  = map nub sequenced
+    removeNull   = filter (not . null) removeDupes
+    sortAllLists = map sort removeNull
+    sortByLength = sortBy (comparing length) sortAllLists
+
+removeSupersets :: Eq a => [[a]] -> [[a]]
+removeSupersets s = filter (not . null) result
+  where
+    prev n  = take n s
+    check x = checkForSupersets (s!!x) (prev x)
+    result  = map (\x -> if check x
+                        then []
+                        else s!!x
+                 )[0..(length s) - 1]
+
+checkForSupersets :: Eq a => [a] -> [[a]] -> Bool
+checkForSupersets current previous = any (`elem` subs) previous
```

Comment:
Ah, I see where your confusion comes from. In the documentation it says that it is *equivalent* to `isSubsequenceOf x y is equivalent to elem x (subsequences y)` and it is indeed equivalent in the sense that it gives the same result. But it's implemented a lot more efficiently than simply trying all possible subsequences.